File: 0150-evaluate-reverse-polish-notation/0150-evaluate-reverse-polish-notation.cpp
// TC- O(n)
// SC- O(n)

class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        stack<int> st;
        for(int i=0; i<tokens.size(); i++){
            string ch = tokens[i];
            if(ch=="+" or ch=="-" or ch=="*" or ch=="/"){
                int a= st.top();
                st.pop();
                int b= st.top();
                st.pop();
                if(ch=="+")
                    st.push(a+b);
                if(ch=="-")
                    st.push(b-a);
                if(ch=="*")
                    st.push(a*b);
                if(ch=="/")
                    st.push(b/a); 
            }
            else
                st.push(stoi(ch));
        }
        return st.top();
    }
};

File: 0150-evaluate-reverse-polish-notation/0150-evaluate-reverse-polish-notation.py
# TC- O(n)
# SC- O(n)

class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stack = []
        for i in range(len(tokens)):
            ch = tokens[i]
            if ch == "+":
                b = int(stack.pop())
                a = int(stack.pop())
                stack.append(str(a+b))
            
            elif ch == "-":
                b = int(stack.pop())
                a = int(stack.pop())
                stack.append(str(a-b))
            
            elif ch == "*":
                b = int(stack.pop())
                a = int(stack.pop())
                stack.append(str(a*b))
            
            elif ch == "/":
                b = int(stack.pop())
                a = int(stack.pop())
                # print(a, b, int(a/b), "Hellow")
                stack.append(str(int(a/b)))
            
            else:
                stack.append(ch)
            # print(stack[-1], " ")
        return int(stack[-1])
